### PR TITLE
Adds support for legacy kubernetes schemas.

### PR DESCRIPTION
**Pull request by tlm**
In juju 2.9 we changed the schemas supported by Kubernetes as pre 2.9
they where incorrect. This change allows existing credentials to
continue working with Juju 2.9.

It's expected that in juju 3.0+ this support will be dropped.

## Checklist

 - [x] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [x] Comments answer the question of why design decisions were made

## QA steps

Add a kubernetes cluster using Juju 2.8 

juju add-k8s microk8s mytestcluster

Compile this branc and then bootstrap using the added cluster above

juju bootstrap mytestcluster

## Bug reference

https://bugs.launchpad.net/juju/+bug/1918486

### CODE REVIEW

**Comment by tlm** [comment]
!!build!!

**Comment by tlm** [comment]
!!build!!

**Comment by tlm** [comment]
[Add support for older style kubernetes credentials to 2.9 branch](https://trello.com/c/PZG2pRBB/1430-add-support-for-older-style-kubernetes-credentials-to-29-branch)

File: caas/kubernetes/cloud/credential.go [hunk 1]
```diff
@@ -82,13 +82,131 @@ var SupportedCredentialSchemas = map[cloud.AuthType]cloud.CredentialSchema{
...
+	switch cred.AuthType() {
+	case cloud.OAuth2WithCertAuthType:
+		return migrateOAuth2WithCertAuthType(cred)
+	case cloud.CertificateAuthType:
+		return migrateCertificateAuthType(cred)
+	default:
+		return cloud.Credential{}, errors.NotSupportedf(
+			"migration for auth type %s", cred.AuthType())
+	}
+}
```
Comment (wallyworld) [inline caas/kubernetes/cloud/credential.go:169]: nit - can we move this up above the 2 unexported migrate functions
right now, it sits between them

File: cloud/clouds.go [hunk 2]
```diff
@@ -55,13 +55,26 @@ const (
 	// OAuth2AuthType is an authentication type using oauth2.
 	OAuth2AuthType AuthType = "oauth2"
 
+	// OAuth2WithCertAuthType is an authentication type using oauth2 and a client certificate
+	// NOTE: This type should never be used in practice as it doesn't exist. We
+	// maintain it here to maintain backwards functionality for Juju in a pre
+	// 2.9 world. Consider using OAuth2AuthType instead.
+	OAuth2WithCertAuthType AuthType = "oauth2withcert"
```
Comment (wallyworld) [inline cloud/clouds.go:62]: Can we group the never use types in a separate const block or even just together in this block

File: state/upgrades.go [hunk 3]
```diff
@@ -481,6 +482,80 @@ func updateLegacyLXDCredentialsOps(st *State, cred cloud.Credential) ([]txn.Op,
...
+// Kubernetes clouds. This is to introduce bug fix changes from 2.9 onwards.
+func UpdateLegacyKubernetesCloudCredentials(st *State) error {
+	ops, err := updateLegacyKubernetesCloudAuthTypesOps(st)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return st.db().RunTransaction(ops)
+}
+
+func updateLegacyKubernetesCloudAuthTypesOps(st *State) ([]txn.Op, error) {
```
Comment (wallyworld) [inline state/upgrades.go:495]: this does more than just update the other types
perhaps just call it updateLegacyKubernetesCredentialsOps() and rename then current updateLegacyKubernetesCredentialsOps to updateLegacyKubernetesCredentialValuesOps or something

File: state/upgrades.go [hunk 4]
```diff
@@ -481,6 +482,80 @@ func updateLegacyLXDCredentialsOps(st *State, cred cloud.Credential) ([]txn.Op,
...
+}
+
+func updateLegacyKubernetesCredentialsOps(st *State, cloudName string) ([]txn.Op, error) {
+	coll, closer := st.db().GetRawCollection(cloudCredentialsC)
+	defer closer()
+	iter := coll.Find(bson.M{"cloud": cloudName}).Iter()
+	defer iter.Close()
+
+	var ops []txn.Op
+	var doc cloudCredentialDoc
```
Comment (wallyworld) [inline state/upgrades.go:530]: In upgrade steps, we avoid using the entity doc - we read / write bson maps instead as these are generic abd not affected by any incompatible changes as the code evolved

File: state/upgrades_test.go [hunk 5]
```diff
@@ -4995,6 +4995,229 @@ func (s *upgradesSuite) TestTranslateK8sServiceTypes(c *gc.C) {
...
+	expectedClouds := []bson.M{{
+		"_id":  "localhost",
+		"name": "localhost",
+		"type": "kubernetes",
+		"auth-types": []interface{}{
+			"certificate",
+			"clientcertificate",
+			"oauth2",
+			"oauth2withcert",
+			"userpass",
```
Comment (wallyworld) [inline state/upgrades_test.go:5040]: I don't think we want to include the depecated auth types after upgrade - we need ti support reading them in from an existing yaml file etc but we should not write them out here

**Review by wallyworld — COMMENTED** [review]
A couple of things to discuss

File: caas/kubernetes/cloud/credential.go (hunk 1 above)
Comment (tlm) [reply caas/kubernetes/cloud/credential.go:169]: nah trying to keep things alphabetical ordered in the file which I have now fixed.

File: state/upgrades_test.go (hunk 5 above)
Comment (tlm) [reply state/upgrades_test.go:5040]: Done & Done

File: state/upgrades.go (hunk 3 above)
Comment (tlm) [reply state/upgrades.go:495]: Done & Done

File: cloud/clouds.go (hunk 2 above)
Comment (tlm) [reply cloud/clouds.go:62]: Done & Done

File: state/upgrades.go (hunk 4 above)
Comment (tlm) [reply state/upgrades.go]: Done & Done

File: state/upgrades_test.go [hunk 6]
```diff
@@ -4995,6 +4995,219 @@ func (s *upgradesSuite) TestTranslateK8sServiceTypes(c *gc.C) {
...
+		"owner":          "admin",
+		"cloud":          "localhost",
+		"name":           "streetcred",
+		"revoked":        false,
+		"auth-type":      "clientcertificate",
+		"invalid":        false,
+		"invalid-reason": "",
+		"attributes": bson.M{
+			"ClientCertificateData": "data",
+			"ClientKeyData":         "data",
```
Comment (wallyworld) [inline state/upgrades_test.go:5197]: These expected attributes look to be the same as the ones created before the upgrade step is run.
Can we construct the "before" attributes to be what would be expected in the field so we can confirm the upgrade worked as expected. Also, don't use "data" for 2 different attributes.

File: state/upgrades_test.go [hunk 7]
```diff
@@ -4995,6 +4995,219 @@ func (s *upgradesSuite) TestTranslateK8sServiceTypes(c *gc.C) {
...
+
+	err = cloudCredColl.Insert(bson.M{
+		"_id":       "localhost#admin#streetcred",
+		"owner":     "admin",
+		"cloud":     "localhost",
+		"name":      "streetcred",
+		"revoked":   false,
+		"auth-type": "oauth2withcert",
+		"attributes": map[string]string{
+			"Token": "this-is-my-token",
```
Comment (wallyworld) [inline state/upgrades_test.go:5098]: Same comment as for TestUpdateLegacyKubernetesCloudCredentialsOAuth2Cert

**Review by wallyworld — APPROVED** [review]
Looks good but I think some test tweaks are needed

File: caas/kubernetes/cloud/credential.go [hunk 8]
```diff
@@ -182,3 +222,90 @@ func CredentialFromKubeConfigContext(
...
```
[the remark this answers is not in the document]
Comment (tlm) [reply caas/kubernetes/cloud/credential.go:259]: Done & Done

File: cloud/clouds.go [hunk 9]
```diff
@@ -60,8 +60,16 @@ const (
 	JSONFileAuthType AuthType = "jsonfile"
 
 	// CertificateAuthType is an authentication type using certificates.
+	// NOTE: This type should never be used in practice as it doesn't exist. We
+	// maintain it here to maintain backwards functionality for Juju in a pre
+	// 2.9 world. Consider using ClientCertificateAuthType instead. This note
+	// applies to Kubernetes.
 	CertificateAuthType AuthType = "certificate"
```
[the remark this answers is not in the document]
Comment (tlm) [reply cloud/clouds.go:67]: This is still being used by LXD which is why I left it here.

File: state/upgrades_test.go (hunk 7 above)
Comment (tlm) [reply state/upgrades_test.go:5098]: Done & Done

File: state/upgrades_test.go (hunk 6 above)
Comment (tlm) [reply state/upgrades_test.go:5197]: Done & Done

File: cloud/clouds.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tlm) [reply cloud/clouds.go:67]: Done & Done

**Comment by tlm** [comment]
$$merge$$

**Pull request merged**